.github/workflows/release-reusable-upload.yaml

```diff
@@ -0,0 +1,268 @@
+                FILE_KEY=$(echo $keypair_file | sed "s|^${{ env.BASE_ARTIFACTS_VERSIONED_DIR }}/||")
+                echo "Uploading keypair file key: $CF_KEYPAIRS_BUCKET_ROOT_KEY/$FILE_KEY"
+                aws s3api put-object --bucket ${{ env.CF_BUCKET_NAME }} --key "$CF_KEYPAIRS_BUCKET_ROOT_KEY/$FILE_KEY" --body "$keypair_file" --acl public-read --endpoint-url $ENDPOINT_URL
+              fi
+            done
+          fi
+
+          # Upload the previously created archives to R2
+          CF_ARCHIVES_BUCKET_ROOT_KEY="${{ env.CF_BUCKET_ROOT_KEY }}/${{ inputs.environment }}/${{ env.VERSION }}/archives"
+
+          for file in "${{ env.TAR_ARCHIVE_FILE }}" "${{ env.ZIP_ARCHIVE_FILE }}"; do
+            FILE_NAME=$(basename "${file}")
+            echo "Uploading archive: ${FILE_NAME}"
+            aws s3api put-object --bucket ${{ env.CF_BUCKET_NAME }} --key "$CF_ARCHIVES_BUCKET_ROOT_KEY/$FILE_NAME" --body "$file" --acl public-read --endpoint-url $ENDPOINT_URL
+            aws s3api put-object --bucket ${{ env.CF_BUCKET_NAME }} --key "$CF_ARCHIVES_BUCKET_ROOT_KEY/$FILE_NAME.sha256" --body "$file.sha256" --acl public-read --endpoint-url $ENDPOINT_URL
+          done
+
+          # Convert list to a compacted JSON array output
+          echo program_files=$(echo "${PROGRAM_FILES}" | jq -R . | jq -s . | jq -c .) >> "$GITHUB_OUTPUT"
+
+      - name: Generate artifact manifest
+        run: |
+          echo "Generating artifact manifest..."
+
+          # Read build metadata if it exists
+          BUILD_REF="${{ inputs.ref }}"
+          COMMIT_SHA="${{ github.sha }}"
+          if [ -f "${{ env.BASE_ARTIFACTS_VERSIONED_DIR }}/build-metadata.json" ]; then
+            BUILD_REF=$(jq -r '.ref // "${{ inputs.ref }}"' ${{ env.BASE_ARTIFACTS_VERSIONED_DIR }}/build-metadata.json)
+            COMMIT_SHA=$(jq -r '.commit_sha // "${{ github.sha }}"' ${{ env.BASE_ARTIFACTS_VERSIONED_DIR }}/build-metadata.json)
+          fi
+
+          # Check if any artifacts exist (excluding build-metadata.json)
+          ARTIFACT_COUNT=$(find ${{ env.BASE_ARTIFACTS_VERSIONED_DIR }} \( -name "*.so" -o \( -name "*.json" -a ! -name "build-metadata.json" -a ! -name "manifest.json" \) \) | wc -l)
+          if [ "$ARTIFACT_COUNT" -eq 0 ]; then
+            echo "Error: No artifacts found to include in manifest"
+            exit 1
+          fi
+
+          echo "Found $ARTIFACT_COUNT artifact(s) to include in manifest"
+          echo "Build type: ${{ env.BUILD_TYPE }}"
+
+          cat > ${{ env.BASE_ARTIFACTS_VERSIONED_DIR }}/manifest.json << EOF
+          {
+            "version": "${{ env.VERSION }}",
+            "build_type": "${{ env.BUILD_TYPE }}",
+            "branch": "$BUILD_REF",
+            "ref": "$BUILD_REF",
+            "environment": "${{ inputs.environment }}",
+            "uploaded_at": "$(date -u +%Y-%m-%dT%H:%M:%SZ)",
+            "commit_sha": "$COMMIT_SHA",
+            "github_ref": "${{ github.ref }}",
```

Comment:
do we need this stuff? isn't all this automatically clear from the path of the release (folder/file name should have github ref), R2 metadata (upload time)